service/storage.go

```diff
@@ -0,0 +1,290 @@
+package service
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+)
+
+var ErrExist = errors.New("object already exists")
+
+var storageNumberAssetsPreload = 3000
+
+// storage is an in memory store which holds users, assets and favourites.
+// It is safe to use concurrently.
+type storage struct {
+	usersMu sync.RWMutex // guards users
+	// maps user id to User
+	users       map[uint]*user
+	userCounter uint         // monotonially increasing counter
+	assetsMu    sync.RWMutex // guards assets
+	// maps user id to Asset
+	assets       map[uint]*asset
+	assetCounter uint         // monotonially increasing counter
+	favouritesMu sync.RWMutex // guards favourites
+	// maps user id to a list of asset ids
+	favourites map[uint][]uint
+}
+
+func newStorage() *storage {
+	return &storage{
+		users:      make(map[uint]*user),
+		assets:     make(map[uint]*asset),
+		favourites: make(map[uint][]uint),
+	}
+}
+
+// // range assets and call f for every asset in the storage. if f returns false,
+// // the loop is stopped.
+// func (s *storage) rangeAssets(f func(*asset) bool) {
+// 	s.assetsMu.RLock()
+// 	defer s.assetsMu.RUnlock()
+// 	for _, a := range s.assets {
+// 		if !f(a) {
+// 			return
+// 		}
+
+// 	}
+// }
+
+// // range users and call f for every asset in the storage. if f returns false,
+// // the loop is stopped.
+// func (s *storage) rangeUsers(f func(*user) bool) {
+// 	s.usersMu.RLock()
+// 	defer s.usersMu.RUnlock()
+// 	for _, u := range s.users {
+// 		if !f(u) {
+// 			return
+// 		}
+
+// 	}
+// }
+
+func (s *storage) getUsers() []*user {
+	s.usersMu.RLock()
```

Comment:
Is there any benefit of locking here if a pointer to "shared object" is returned?

Reply:
No, and in fact, this is a terrible mistake. Race conditions could emerge since this code is not thread-safe but was intended to be. I should return `[]user` instead. Can I make a quick patch to the code or should I live it as is?